File: Python/5.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python
#
# [5] Longest Palindromic Substring
#
# https://leetcode.com/problems/longest-palindromic-substring/description/
#
# algorithms
# Medium (28.14%)
# Likes:    4628
# Dislikes: 417
# Total Accepted:    697.5K
# Total Submissions: 2.5M
# Testcase Example:  '"babad"'
#
# Given a string s, find the longest palindromic substring in s. You may assume
# that the maximum length of s is 1000.
# 
# Example 1:
# 
# 
# Input: "babad"
# Output: "bab"
# Note: "aba" is also a valid answer.
# 
# 
# Example 2:
# 
# 
# Input: "cbbd"
# Output: "bb"
# 
# 
#

# @lc code=start
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if len(s) == 1:
            return s
            
        n = len(s)
        dp = [[False for _ in range(n)] for _ in range(n)]
        
        for i in range(n):
            dp[i][i] = True
        for i in range(n-1):
            if s[i] == s[i+1]:
                dp[i][i+1] = True

        for l in range(3, n+1):
            for i in range(n-l+1):
                j = i + l - 1
                if s[i] == s[j]:
                    dp[i][j] = dp[i+1][j-1]

        ans = ""
        for i in range(n):
            for j in range(n):
                if dp[i][j] and j - i + 1 > len(ans):
                    ans = s[i:j+1]
        return ans
                    

        # 扩展中心法
        # 遍历所有的回文串中心（考虑奇数、偶数长度），从中心向外扩展可以用O(n)的时间找到该中心位置对应的最长回文串，这个思路其实相当直接，但时空复杂度相对于上面两个动规反而更好

        res = ''
        for i in range(len(s)):
            # odd case
            tmp = self.expand(s, i, i)
            if len(tmp) > len(res):
                res = tmp 
            
            # even case
            tmp = self.expand(s, i, i+1)
            if len(tmp) > len(res):
                res = tmp 
        return res
    
    def expand(self, s, l, r):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        return s[l+1: r]
# @lc code=end

sol = Solution()
s = 'cbbd'
print(sol.longestPalindrome(s))

File: C++/5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode.cn id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */

#include <string>
#include<vector>

using namespace std; 

// @lc code=start
class Solution {
public:
    string longestPalindrome(string s) {
        // const int size_s = s.size();

        // int max_s = 0, max_l = 0;
        // for (int i=0; i<size_s;) {
        //     int start = i, end = i;
            
        //     // Skip duplicate characters.
        //     // don't need to check whether the palindrome is odd or even compared to many brute force solutions.
        //     while (end+1<size_s and s[end]==s[end+1]) end ++;
        //     i = end + 1;

        //     while (start-1>=0 and end+1<size_s and s[start-1]==s[end+1]) {
        //         start--;
        //         end++;
        //     }
            
        //     if (end-start+1 > max_l) {
        //         max_l = end-start+1;
        //         max_s = start;
        //     }
        // }
        // return s.substr(max_s, max_l);
        if (s.empty()) return "";
        
        int n=s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        for (int i=0; i<n; ++i) dp[i][i] = true;
        
        int left = 0, len = 1;
        for (int i=0; i<n; ++i) {
            for (int j=0; j<i; ++j) {
                dp[j][i] = (s[i]==s[j] && (i-j<2 || dp[j+1][i-1]));
                
                if (dp[j][i] && len<i-j+1) {
                    len = i-j+1;
                    left = j;
                }
            }
        }
        return s.substr(left, len);

        // 
        int n = s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        for (int i=0; i<n; ++i) dp[i][i] = true;
        
        int max_len = 1, start = 0;
        for (int len=2; len<=n; ++len) {
            for (int i=0; i<n-len+1; ++i) {
                int j = i + len - 1;
                
                if (s[i] == s[j] && (j-i==1 || dp[i+1][j-1])) 
                    dp[i][j] = true;
                
                
                if (dp[i][j] && len > max_len) {
                    max_len = len;
                    start = i;
                }
            }
        }
        return s.substr(start, max_len);
        

        // 

        int start = 0, max_len = 0;
        for (int i=0; i<s.size(); ++i) {
            helper(s, i, i, start, max_len);
            helper(s, i, i+1, start, max_len);
        }
        return s.substr(start, max_len);
    }
    
    void helper(string& s, int left, int right, int& start, int& max_len) {
        while (left>=0 && right<s.size() && s[left]==s[right]) {
            left -= 1;
            right += 1;
        }
        if (max_len < right-left-1) {
            start = left + 1;
            max_len = right -left - 1;
        }
    }
};
// @lc code=end

